File: src/main/java/me/pcasaes/hexoids/infrastructure/kafka/DelayedStartConsumerRebalanceListener.java
package me.pcasaes.hexoids.infrastructure.kafka;

import io.smallrye.reactive.messaging.kafka.KafkaConsumerRebalanceListener;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.inject.Named;
import org.apache.kafka.clients.consumer.Consumer;

import java.util.Collection;

@ApplicationScoped
@Named("DelayedStartConsumerRebalanceListener")
public class DelayedStartConsumerRebalanceListener implements KafkaConsumerRebalanceListener {

    private final DelayedStartConsumerHandler handler;

    @Inject
    public DelayedStartConsumerRebalanceListener(DelayedStartConsumerHandler handler) {
        this.handler = handler;
    }

    @Override
    public void onPartitionsAssigned(Consumer<?, ?> consumer, Collection<org.apache.kafka.common.TopicPartition> partitions) {
        this.handler
                .onStarted()
                .await()
                .indefinitely();
    }

}

File: src/main/kotlin/me/pcasaes/hexoids/infrastructure/kafka/DelayedStartConsumerRebalanceListener.kt
package me.pcasaes.hexoids.infrastructure.kafka

import io.smallrye.reactive.messaging.kafka.KafkaConsumerRebalanceListener
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.inject.Named
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.common.TopicPartition

@ApplicationScoped
@Named("DelayedStartConsumerRebalanceListener")
class DelayedStartConsumerRebalanceListener @Inject constructor(
    private val handler: DelayedStartConsumerHandler
) : KafkaConsumerRebalanceListener {

    override fun onPartitionsAssigned(consumer: Consumer<*, *>, partitions: MutableCollection<TopicPartition>) {
        this.handler
            .onStarted()
            .await()
            .indefinitely()
    }
}
